lib/src/rules/unnecessary_parenthesis.dart

```diff
@@ -68,6 +68,12 @@ class _Visitor extends SimpleAstVisitor<void> {
       return;
     }
 
+    // https://github.com/dart-lang/linter/issues/2944
+    // (() => foo()).bar();
+    if (expression is FunctionExpression && parent is MethodInvocation) {
```

Comment:
In addition to MethodInvocation this same problem may happen for (deep breath):

* getter access: `(() => foo()).bar;`
* setter access: `(() => foo()).bar = 7;`
* operators: `(() => foo())[6];` or `(() => foo())++;`
* type instantiation: `(() => foo()).bar<int>;`

Reply:
Right!  Good call.  I'll update.